File: user_manage.lua
Databases = {}

function ValidateUsername()
  while true do
    io.write("\n" .. "Enter username: ")
    local user = io.read()
    if string.find(user, "postgres") then
      io.stderr:write("ERROR: username cannot contain 'postgres'")
    else
      return string.lower(user)
    end
  end
end

function ValidatePassword()
  while true do
    io.write("Enter password: ")
    local password = io.read()
    if #password < 9 then
      io.stderr:write("ERROR: password too short\n")
    else
      return password
    end
  end
end

function VerifyResult(user, file_name)
  local file = assert(io.open(file_name, "r"))
  local content = file:read("*all")
  print("\n" .. content)
  file:close()
  io.write("Is this OK? (y/n) ")
  while true do
    local input = io.read()
    if input == "y" then
      print(string.format("Created .sql script for %s.\n", user))
      break
    elseif input == "n" then
      os.remove(file_name)
      print(string.format("Discarded .sql script for %s.\n", user))
      return
    else
      print("Only 'y' or 'n' is accepted.")
      input = io.read()
    end
  end

  io.write("Apply? (y/n) ")
  local command = string.format("psql -U postgres -f %s", file_name)
  while true do
    local input = io.read()
    if input == "y" then
      print()
      os.execute(command)
      print()
      break
    elseif input == "n" then
      os.remove(file_name)
      print(string.format("Discarded .sql script for %s.\n", user))
      break
    else
      print("Only 'y' or 'n' is accepted.")
    end
  end
end

function CreateUser(databases)
  local user = ValidateUsername()
  local password = ValidatePassword()
  local file_name = string.format("%s.sql", user)
  local file = assert(io.open(file_name, "w"))
  file:write(string.format("CREATE USER %s WITH PASSWORD '%s';\n", user, password))
  for _, db in ipairs(databases) do
    file:write(string.format("GRANT CONNECT ON DATABASE %s TO %s;\n", db, user))
  end
  for _, db in ipairs(databases) do
    file:write(string.format("\\c %s\n", db))
    file:write(string.format("GRANT USAGE ON SCHEMA public TO %s;\n", user))
    file:write(string.format("GRANT SELECT ON ALL TABLES IN SCHEMA public TO %s;\n", user))
  end
  file:close()
  VerifyResult(user, file_name)
end

function CreateUserRW(databases)
  local user = ValidateUsername()
  local password = ValidatePassword()
  local file_name = string.format("%s.sql", user)
  local file = assert(io.open(file_name, "w"))
  file:write(string.format("CREATE USER %s WITH PASSWORD '%s';\n", user, password))
  for _, db in ipairs(databases) do
    file:write(string.format("GRANT CONNECT ON DATABASE %s TO %s;\n", db, user))
  end
  for _, db in ipairs(databases) do
    file:write(string.format("\\c %s\n", db))
    file:write(string.format("GRANT USAGE ON SCHEMA public TO %s;\n", user))
    file:write(string.format("GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA public TO %s;\n", user))
  end
  file:close()
  VerifyResult(user, file_name)
end

function RevokeUser(databases)
  local user = ValidateUsername()
  local file_name = string.format("revoke_%s.sql", user)
  local file = assert(io.open(file_name, "w"))
  for _, db in ipairs(databases) do
    file:write(string.format("REVOKE ALL ON DATABASE %s FROM %s;\n", db, user))
  end
  file:write(string.format("REVOKE ALL ON SCHEMA public FROM %s;\n", user))
  file:write(string.format("DROP USER %s;\n", user))
  file:close()
  VerifyResult(user, file_name)
end

function ChangePassword()
  local user = ValidateUsername()
  local password = ValidatePassword()
  local file_name = string.format("chpw_%s.sql", user)
  local file = assert(io.open(file_name, "w"))
  file:write(string.format("ALTER USER %s WITH PASSWORD '%s';", user, password))
  file:close()
  VerifyResult(user, file_name)
end

function ValidateEnvVars()
  local errors = 0
  if not os.getenv("PGPASSWORD") then
    io.stderr:write("PGPASSWORD environment variable not defined.\n")
    errors = errors + 1
  end
  if not os.getenv("PGHOST") then
    io.stderr:write("PGHOST environment variable not defined.\n")
    errors = errors + 1
  end
  return errors
end

function Main()
  local env_var_errors = ValidateEnvVars()
  if env_var_errors > 0 then
    os.exit(1)
  end
  while true do
    local options = {
      ["1"] = function() CreateUser(Databases) end,
      ["2"] = function() CreateUserRW(Databases) end,
      ["3"] = function() RevokeUser(Databases) end,
      ["4"] = function() ChangePassword() end,
      ["q"] = function() print("Exiting...") end,
    }
    io.write("Choose action:\n[1] Create user\n[2] Create RW user\n[3] Revoke user\n[4] Change password\n[q] Exit\n")
    io.write("Selected option: ")
    local user_input = io.read()
    local selected = options[user_input]
    if selected ~= nil then
      selected()
      if user_input == "q" then
        break
      end
    else
      print("Unsupported input.")
    end
  end
  os.execute("rm -f *.sql")
end

Main()

File: user_manage.hs
import Control.Monad (when)
import Control.Monad.Reader (ask, liftIO, runReaderT, ReaderT)
import Data.Char (toLower)
import Data.Foldable (traverse_)
import Data.List (isInfixOf, intercalate)
import Data.Maybe (isNothing)
import System.Directory (removeFile)
import System.Environment (lookupEnv)
import System.Exit (exitSuccess, die)
import System.IO (hFlush, stdout)
import Text.Printf (printf)

newtype Username = User { getUser :: String }
newtype Password = Password { getPassword :: String }
newtype DatabaseList = DBList { getDBList :: [String] }
type SqlStatement = String

main :: IO ()
main = do
  envVars <- runReaderT parseEnvVars inputsEnvVar
  when (any (isNothing . snd) envVars) $ do
    let missingVars = fst <$> filter (isNothing . snd) envVars
    let errorMessage = unlines $ (++ " environment variable not defined") <$> missingVars
    die errorMessage
  displayMessage "Enter username: "
  user <- promptUser
  when ("postgres" `isInfixOf` getUser user) $
   die "Username cannot contain postgres"
  displayMessage "Enter password: "
  password <- promptPassword
  when (length (getPassword password) < 9) $
    die "Password too short"
  promptAction user password inputsDB

displayMessage :: String -> IO ()
displayMessage message = do
  putStr message
  hFlush stdout

inputsDB :: DatabaseList
inputsDB = DBList ["web", "web-api"]

inputsEnvVar :: [String]
inputsEnvVar = ["PGHOST", "PGPASSWORD"]

promptUser :: IO Username
promptUser = User <$> getLine

promptPassword :: IO Password
promptPassword = Password <$> getLine

promptAction :: Username -> Password -> DatabaseList -> IO ()
promptAction username password dbs = do
  putStrLn "Choose action:"
  putStrLn "[1] Create username"
  putStrLn "[2] Revoke username"
  putStrLn "[3] Change password"
  putStrLn "[q] Exit\n"
  putStr "Selected option: "
  action <- getLine
  case action of
    "1" -> do
      writeFile file_name $
        userCreateSQL username password dbs
      result <- verifyResult username file_name
      print result
      where
        file_name = getUser username ++ ".sql"
    "2" -> do
      writeFile file_name $
        userRevokeSQL username dbs
      result <- verifyResult username file_name
      print result
      where
        file_name = "revoke_" ++ getUser username ++ ".sql"
    "3" -> do
      writeFile file_name $
        userPasswordChangeSQL username password
      result <- verifyResult username file_name
      print result
      where
        file_name = "chpw_" ++ getUser username ++ ".sql"
    "q" -> do
      putStrLn "Exiting...\n"
      exitSuccess
    _ -> do
      putStrLn "Unsupported input\n"
      promptAction username password dbs

verifyResult :: Username -> FilePath -> IO Bool
verifyResult username file_name = do
  putStrLn ""
  content <- readFile file_name
  putStrLn content
  putStrLn "Is this OK? (y/n)"
  answer <- getLine
  case answer of
    "y" -> do
      undefined
      pure True -- should proceed to applyResult here but no idea how
      -- I don't feel like supplying more arguments just for sake of this
      -- while it should be own FilePath -> IO () function
    "n" -> do
      removeFile file_name
      putStrLn $ "Discarded .sql script for " ++ getUser username
      pure False
    _ -> do
      putStrLn "Only y/n are accepted."
      verifyResult username file_name

parseEnvVars :: ReaderT [String] IO [(String, Maybe String)]
parseEnvVars = do
  envVars <- ask
  liftIO $
    traverse (
      \var -> do
        result <- lookupEnv var
        pure (var, result)
        ) envVars

userCreateSQL :: Username -> Password -> DatabaseList -> SqlStatement
userCreateSQL username password dbs =
  unlines $
    [createUserStatement]
      ++ (mkConnectGrant <$> getDBList dbs)
      ++ (mkUsageGrant <$> getDBList dbs)
  where
    createUserStatement = printf "CREATE USER %s WITH PASSWORD '%s';" (getUser username) (getPassword password)
    mkConnectGrant db = printf "GRANT CONNECT ON DATABASE %s TO %s;" db (getUser username)
    mkUsageGrant db =
      printf "\\c %s\n" db
        ++ "\n"
        ++ printf "GRANT USAGE ON SCHEMA public TO %s;" (getUser username)
        ++ "\n"
        ++ printf "GRANT SELECT ON ALL TABLES IN SCHEMA public TO %s;" (getUser username)

userRevokeSQL :: Username -> DatabaseList -> SqlStatement
userRevokeSQL username dbs = do
  unlines $
    mkRevokeGrant <$> getDBList dbs
      ++ [revokeSchemaPublic]
      ++ [dropUserStatement]
  where
    mkRevokeGrant db = printf "REVOKE ALL ON DATABASE %s FROM %s;\n" db (getUser username)
    revokeSchemaPublic = printf "REVOKE ALL ON SCHEMA public FROM %s;\n" (getUser username)
    dropUserStatement = printf "DROP USER %s;\n" (getUser username)

userPasswordChangeSQL :: Username -> Password -> SqlStatement
userPasswordChangeSQL username password = do
  unlines [userPasswordChangeSQLStatement (getUser username) (getPassword password)]
  where
    userPasswordChangeSQLStatement = printf "ALTER USER %s WITH PASSWORD '%s';"
